Glossary (#361)
ArcadeAI/docs: pages/home/glossary.mdx

---
title: "Glossary"
description: "Glossary of terms used in the Arcade platform"
---

# Arcade Glossary

## Agents and Tools

```mermaid
graph TD
    A[Agent] --> B[Toolkit]
    B --> C[Tool]
```

### Agent

An 'agent' is the application you are building.  It can be a chatbot, a web application, a mobile app, or any other type of application that happens to use an LLM to as part of its functionality.  Agents interact with the world by calling tools.  Helping you build, test, authenticate, and deploy tools is what Arcade is all about.

### Tool

A 'tool' is a function that can be called by an agent which interacts with another service - commonly via an API, filesystem, database etc.  Tools are written in python and deployed by running a worker which contains the toolkit's code.  Tools are defined by the `@tool()` decorator and will be passed `ToolContext` as the first argument.  If a tool has dependencies that are not met (a secret is not provided, for example), the tool will fail to execute.

An example of a tool, scoped by it's toolkit would be `Gmail.SendEmail`.

*Learn more about [tools](/home/build-tools/create-a-toolkit).*

> ### Toolkit

Reviewer: Kinda feels like we should define this before 'tool'. Agent->toolkit->tool

Author: 91236394b1d4a9db19a70160745b417509c64b13

### Project

A 'project' is a collection of agents and tools.  It is the lowest level of organization in Arcade.  You can have multiple projects, and each project can have multiple agents and tools.  Different Accounts can be members of different projects, and each project will have different API keys.

### API Key

An 'API key' is a secret key that is used to authenticate requests to the Arcade API.  It is used to identify the project that the request is for.  API keys are project-specific.

## Authentication and Billing

```mermaid
graph TD
    A[Arcade] --> B[Agent]
    B --> C[User]
```

### User

A 'user' is your end-user, the person who is using your application or agent.  Users are counted by the unique `user_id` properties sent when calling tools.  `user_id` values are commonly email addresses, but can be any string or number.

### Monthly Active Users (MAU)

Monthly Active Users are the unique end-users (counted by `user_id`) who have executed a tool in your app/agent within the past month.  If the same `user_id` calls a tool multiple times in the same month, or executes multiple tools, it is only counted once.

### Authentication Starts (Auth Starts)

> Authentication Starts are the number of unique users (by `user_id`) who have started the authentication flow for a specific toolkit.  This is a proxy for the number of new users who are using your agent.  The same user authenticating to multiple toolkits will be counted multiple times (e.g. once for Slack and once for Google).

Reviewer 2: Since multiple toolkit can use the same auth provider (think Gmail, GoogleCalendar, etc.)
```suggestion
Authentication Starts are the number of unique users (by `user_id`) who have started the authentication flow for a specific auth provider.  This is a proxy for the number of new users who are using your agent.  The same user authenticating to multiple toolkits will be counted multiple times (e.g. once for Slack and once for Google).
```

Reviewer 2: Also a question for myself - If my user "auth starts" for Gmail's `read` scope, and then later "auth starts" for Gmail's `write` scope, then is this counted as 1 or 2 auth starts?

Author: Clarified via d28d5fdda4a0ba52031c481e9eda23676db6864b

### Tool Executions

> A 'tool execution' is a single call to a tool to interact with a remote system or service. The tool execution itself may fail (e.g. the user does not have permission to call the tool), but as long as the execution was able to be routed to a worker, the tool execution will be counted.

Reviewer 2: In my opinion it seems unfair to charge a customer for a tool execution if one of our official tools fail because of a `KeyError`. This would be our fault and not theirs. For now though we don't have the ability determine if the tool execution failure is ours or not

*Learn more about [tool executions](/home/use-tools/tools-overview).*

### Standard and Pro Tool executions

Arcade tools are divided into 2 categories: Standard and Pro.  While all tools have some cost for Arcade to run, Pro tools are significantly more costly - either due to infrastructure costs, the complexity of the tool, or a cost imposed by the provider of the tool.  Pro tools cost more to execute and have different limits.

Learn more about tool pricing [here](https://www.arcade.dev/pricing).

### Bring Your Own Credentials (BYOC)

> BYOC is a feature that allows you to use your own credentials to certain premium tools.  This changes the cost of the tool execution, as you will be charged directly by the provider of the tool, rather than relying on Arcade to pay the bill for you.

Author: fixed in 646bc80d37d85a95149305c2eace3118fc765006

*Learn more about [tool context](/home/build-tools/tool-context).*

### MCP (Model Context Protocol)

MCP is an open standard protocol that enables LLMs to access and use tools and data sources. Arcade tools go further than MCP with enterprise-grade authentication, secure token management, and fine-grained permissions. This allows your tools to be accessible to any LLM that supports MCP, while maintaining security and user privacy.

Arcade Tools can be presented as an MCP server (via the Arcade Engine), allowing any LLM that supports MCP to access your tools.  In the future, remote MCP servers will be able to be added to your project as additional workers, making their tools available to your agents.

*Learn more about [MCP and Arcade](/home/mcp-overview).*

### OXP

OXP is Arcade's proprietary protocol for tool definition and tool calling that enables AI/LLM agents to access tools and services.  OXP is a more secure and performant alternative to MCP, and is the preferred way to access tools from your agents.

## The Arcade Platform

```mermaid
graph TD
    A[Tenant] --> B[Project]
    B --> C[API Key]
    C --> D[Agent]
    Z[Account] --> B
    Z[Account] --> A
```

### Account

> The 'account' is you, the user of Arcade who is developing an application or agent.  You can sign into the Arcade dashboard, manage projects, and more.

Reviewer: ```suggestion
The 'account' is you (or your teammates), the developer(s) whoare using Arcade to build an application or agent.  You can sign into the Arcade dashboard, manage projects, and more.
```

### Tenant

> A 'tenant' is a collection of projects and users.  It is the top-level organization in Arcade.  You can be a member of one or more tenants, and each tenant can have multiple projects.  Billing is handled at the tenant level.

Reviewer: ```suggestion
A 'tenant' is a collection of projects with unified billing details.  It is the top-level unit of organization in Arcade.  You can be a member of one or more tenants, and each tenant can have multiple projects.
```

Learn more about tool pricing [here](https://www.arcade.dev/pricing).

### Bring Your Own Credentials (BYOC)

Bring Your Own Credentials (BYOC) is a feature that allows you to use your own credentials to certain pro tools.  This changes the cost of the tool execution, as you will be charged directly by the provider of the tool, rather than relying on Arcade to pay the bill for you.

To set your own credentials, set the requisite secret within the Arcade Dashboard, overwriting the default 'static' credentials.

### Tool Execution Flow

```mermaid
graph TD
    A[Agent] --> B[Arcade Client]
    B --> C[Arcade Engine]
    C --> D[Worker]
```

### Arcade Client

The 'Arcade client' is the SDK that you use to interact with the Arcade platform.  It is how your agent lists and calls tools, tied back to your project via an API key.  We offer clients for many popular languages and frameworks, including Python, JavaScript, and more.

> *Learn more about [the Arcade clients](/home/arcade-clients).*

Reviewer: nit: need to define TDK / Tool Development Kit too! since that is separate from Client SDK now

### Arcade Engine

> The 'Arcade engine' is the core of the Arcade platform.  It is responsible for routing tool execution requests to the correct worker, and for managing the lifecycle of tool executions.  It is also responsible for checking the user's permissions to call a tool, and for injecting secrets and other information into the tool call.

Reviewer: ```suggestion
The Arcade Engine is the core of the Arcade platform.  It is responsible for routing tool execution requests to the correct worker, managing the lifecycle of tool executions, and for enforcing security and authorization decisions.
```

The Arcade Engine is also responsible for the OAuth flow for your agent's users.  This includes granting and elevating permissions and keeping tokens fresh.

### Worker

> A 'worker' is a process that runs all the tools with a toolkit.  Workers are microservices that interact with the Arcade Engine to respond to tool execution requests.  Workers are either deployed to Arcade's infrastructure via `arcade deploy` or run in your own infrastructure as a 'hybrid worker'.

Reviewer: ```suggestion
A 'worker' is a process that runs all the tools with a toolkit.  Workers are microservices that are called by the Arcade Engine to handle tool execution requests.  Workers are either [deployed to Arcade's infrastructure](https://docs.arcade.dev/home/serve-tools/arcade-deploy) via `arcade deploy`, or run in your own infrastructure as a 'hybrid worker'.
```

---
title: "Glossary"
description: "Glossary of terms used in the Arcade platform"
---

# Arcade Glossary

## Agents and Tools

```mermaid
graph TD
    A[Agent] --> B[Toolkit]
    B --> C[Tool]
```

### Agent

An 'agent' is the application you are building.  It can be a chatbot, a web application, a mobile app, or any other type of application that happens to use an LLM to as part of its functionality.  Agents interact with the world by calling tools.  Helping you build, test, authenticate, and deploy tools is what Arcade is all about.

### Toolkit

A 'toolkit' is a collection of tools that can be used by an agent, grouped logically together by a common theme or provider.  Toolkits are the unit of deployment for tools within Arcade.

### Tool

> A 'tool' is a function that can be called by an agent which performs some action - commonly via an API, filesystem, database, etc.  Tools are written in python and deployed by running a worker which contains the toolkit's code.  Tools are defined by the `@tool()` decorator and will be passed `ToolContext` as the first argument.  If a tool has dependencies that are not met (a secret is not provided, for example), the tool will fail to execute.

Reviewer 3: ```suggestion
A 'tool' is a function that can be called by an agent which performs some action - commonly via an API, filesystem, database, etc.  Tools are written in Python and deployed by running a worker which contains the toolkit's code.  Tools are defined by the `@tool()` decorator and will be passed `ToolContext` as the first argument.  If a tool has dependencies that are not met (a secret is not provided, for example), the tool will fail to execute.
```

Tools are commonly referred to by a qualified name that includes their toolkit. For example, [Gmail.SendEmail](https://docs.arcade.dev/toolkits/productivity/gmail#gmailsendemail)

*Learn more about [tools](/home/build-tools/create-a-toolkit).*

### Tool Context

'Tool context' is an object that is passed to a tool as a parameter when the tool is executed.  It contains information about the tool call, the user for which the tool is being called, and any secrets the tool requires to run.

*Learn more about [tool context](/home/build-tools/tool-context).*

### MCP (Model Context Protocol)

MCP is an open standard protocol that enables LLMs to access and use tools and data sources. Arcade tools go further than MCP with enterprise-grade authentication, secure token management, and fine-grained permissions. This allows your tools to be accessible to any LLM that supports MCP, while maintaining security and user privacy.

> Arcade Tools can be presented as an MCP server (via the Arcade Engine), allowing any LLM that supports MCP to access your tools.  In the future, remote MCP servers will be able to be added to your project as additional workers, making their tools available to your agents.

Reviewer 3: ```suggestion
Arcade Tools can be presented as an MCP server (via the Arcade Engine), allowing any LLM that supports MCP to access your tools.  In the future, you will be able to add remote MCP servers to your project as additional workers, making their tools available to your agents.
```

*Learn more about [MCP and Arcade](/home/mcp-overview).*

### OXP

> OXP is Arcade's proprietary protocol for tool definition and tool calling that enables AI/LLM agents to access tools and services.  OXP is a more secure and performant alternative to MCP, and is the preferred way to access tools from your agents.

Reviewer 3: Is this proprietary? If so, let's consider renaming it so people that already know about https://github.com/OpenExecProtocol/Specification don't complain

## The Arcade Platform

```mermaid
graph TD
    A[Tenant] --> B[Project]
    B --> C[API Key]
    C --> D[Agent]
    Z[Account] --> B
    Z[Account] --> A
```

### Account

> The 'account' is you (or your teammates), the developer(s) whoare using Arcade to build an application or agent.  You can sign into the Arcade dashboard, manage projects, and more.

Reviewer 3: ```suggestion
The 'account' is you (or your teammates), the developer(s) who are using Arcade to build an application or agent. You can sign into the Arcade dashboard, manage projects, and more.
```
